qutip/solver/heom/bofin_fit.py

```diff
@@ -0,0 +1,1011 @@
+"""
+This module provides utilities for fitting bosonic baths through
+the correlation function or spectral density, the fit returns a
+HEOM bath object see the ``qutip.solver.heom.bofin_baths``.
+
+The number of modes for the fit can be indicated by the user or
+determined by requiring a normalized root mean squared error for
+the fit.
+"""
+
+import numpy as np
+from time import time
+try:
+    from mpmath import mp
+    _mpmath_available = True
+except ModuleNotFoundError:
+    _mpmath_available = False
+from scipy.optimize import curve_fit
+from qutip.solver.heom import UnderDampedBath, BosonicBath
+from scipy.interpolate import InterpolatedUnivariateSpline
+
+
+__all__ = ["SpectralFitter", "CorrelationFitter", "OhmicBath"]
+
+
+class SpectralFitter:
+    """
+    A helper class for constructing a Bosonic bath from a fit of the spectral
+    density with a sum of underdamped modes.
+
+    Parameters
+    ----------
+    Q : Qobj
+        Operator describing the coupling between system and bath.
+
+    T : float
+        Bath temperature.
+
+    w : np.array
+        The range on which to perform the fit, it is recommended that it covers
+        at least twice the cutoff frequency of the desired spectral density.
+
+    J : np.array or callable
+        The spectral density to be fitted as an array or function.
+    """
+
+    def __init__(self, T, Q, w, J):
+        self.Q = Q
+        self.T = T
+        self.set_spectral_density(w, J)
+
+    def set_spectral_density(self, w, J):
+        """
+        Sets the spectral density to be fitted. It may be provided either as an
+        array of function values or as a python function. For internal reasons,
+        it will then be interpolated or discretized as necessary.
+        """
+
+        if callable(J):
+            self._w = w
+            self._J_array = J(w)
+            self._J_fun = J
+        else:
+            self._w = w
+            self._J_array = J
+            self._J_fun = InterpolatedUnivariateSpline(w, J)
+
+    @staticmethod
+    def _meier_tannor_SD(w, a, b, c):
+        r"""
+        Underdamped spectral density used for fitting in Meier-Tannor form
+        (see Eq. 38 in the BoFiN paper, DOI: 10.1103/PhysRevResearch.5.013181)
+
+        $J(\omega) = \sum_{i=1}^{k} \frac{2 \alpha_{i}^{2} \Gamma_{i} \omega
+        }{\left( \left( \omega + \Omega_{i}\right)^{2} + \Gamma_{i}^{2}
+        \right)+\left( \omega - \Omega_{i}\right)^{2} + \Gamma_{i}^{2}
+        \right)}$
```

Comment:
This is not the formula used in the code (coupling constant not squared).
Also, we should call the parameters either a,b,c or alpha,Gamma,Omega.
(However, I don't really think we need the equation here, since the method is not public...)